File: server-laravel-single/app/Models/Project.php
<?php

namespace App\Models;

use Rhino\Models\RhinoModel;
use Rhino\Traits\HasAuditTrail;

/**
 * Project — top-level user-owned model in the single-tenant variant.
 *
 * Ownership is by `user_id`. On create, user_id is auto-stamped from the
 * authenticated user. Read/write isolation is enforced by ProjectScope
 * (app/Models/Scopes/ProjectScope.php), which constrains every query to the
 * current user's rows.
 */
class Project extends RhinoModel
{
    use HasAuditTrail;

    protected static function booted(): void
    {
        static::creating(function (Project $project) {
            // Auto-stamp ownership from the authenticated user. Gated on
            // auth()->check() so console/seeder creates (which set user_id
            // explicitly and have no auth user) are left untouched.
            if (auth()->check() && !$project->user_id) {
                $project->user_id = auth()->id();
            }
        });
    }

    protected $fillable = [
        'user_id',
        'title',
        'description',
        'status',
        'budget',
        'internal_notes',
        'starts_at',
        'ends_at',
    ];

    protected $validationRules = [
        'title' => 'required|string|max:255',
        'description' => 'nullable|string',
        'status' => 'required|string|max:255',
        'budget' => 'nullable|numeric',
        'internal_notes' => 'nullable|string',
        'starts_at' => 'nullable|date',
        'ends_at' => 'nullable|date',
    ];

    public static $allowedFilters = ['title', 'status'];
    public static $allowedSorts = ['title', 'status', 'starts_at', 'ends_at'];
    public static $defaultSort = 'created_at';
    public static $allowedFields = [
        'id', 'user_id', 'title', 'description', 'status', 'budget',
        'internal_notes', 'starts_at', 'ends_at', 'created_at', 'updated_at',
    ];
    public static $allowedIncludes = ['tasks', 'user'];

    public function tasks(): \Illuminate\Database\Eloquent\Relations\HasMany
    {
        return $this->hasMany(Task::class);
    }

    public function user(): \Illuminate\Database\Eloquent\Relations\BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}

File: server-rails-single/app/models/project.rb
# frozen_string_literal: true

# Project — top-level user-owned model in the single-tenant variant.
#
# Ownership is by user_id. On create, user_id is auto-stamped from the
# authenticated user (RequestStore). Read/write isolation is enforced by
# Scopes::ProjectScope, which constrains every query to the current user's rows.
class Project < Rhino::RhinoModel
  include Rhino::HasAuditTrail
  include Discard::Model

  rhino_filters :title, :status
  rhino_sorts :title, :status, :starts_at, :ends_at
  rhino_default_sort "created_at"
  rhino_fields :id, :user_id, :title, :description, :status, :budget, :internal_notes, :starts_at, :ends_at, :created_at, :updated_at

  validates :title, length: { maximum: 255 }, allow_nil: true
  validates :status, inclusion: { in: %w[draft active completed archived] }, allow_nil: true

  belongs_to :user
  has_many :tasks, dependent: :destroy

  # Auto-stamp ownership from the authenticated user on create. Skipped when
  # user_id is already set (seeds) or when there is no request user (console).
  before_validation :auto_set_user_id, on: :create

  private

  def auto_set_user_id
    return if user_id.present?
    return unless defined?(RequestStore)

    current_user = RequestStore.store[:rhino_current_user]
    self.user_id = current_user.id if current_user
  end
end
